cuda/base/kernel_launch.cuh

```diff
@@ -0,0 +1,243 @@
+struct device_unpack_2d_impl {
+    using type = T;
+    static __device__ __forceinline__ type unpack(T param, size_type)
+    {
+        return param;
+    }
+};
+
+template <typename ValueType>
+struct device_unpack_2d_impl<compact_dense_wrapper<ValueType>> {
+    using type = matrix_accessor<ValueType>;
+    static __device__ __forceinline__ type
+    unpack(compact_dense_wrapper<ValueType> param, size_type num_cols)
+    {
+        return {param.data, num_cols};
+    }
+};
+
+
+template <typename KernelFunction, typename... KernelArgs>
+__global__ __launch_bounds__(default_block_size) void generic_kernel_1d(
+    size_type size, KernelFunction fn, KernelArgs... args)
+{
+    auto tidx = thread::get_thread_id_flat();
+    if (tidx >= size) {
+        return;
+    }
+    fn(tidx, args...);
+}
+
+
+template <typename KernelFunction, typename... KernelArgs>
+__global__ __launch_bounds__(default_block_size) void generic_kernel_2d(
+    size_type rows, size_type cols, KernelFunction fn, KernelArgs... args)
+{
+    auto tidx = thread::get_thread_id_flat();
+    auto col = tidx % cols;
+    auto row = tidx / cols;
+    if (row >= rows) {
+        return;
+    }
+    fn(row, col, device_unpack_2d_impl<KernelArgs>::unpack(args, cols)...);
+}
+
+
+}  // namespace cuda
+}  // namespace kernels
+
+
+template <typename ValueType>
+kernels::cuda::compact_dense_wrapper<kernels::cuda::cuda_type<ValueType>>
+compact(matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {kernels::cuda::as_cuda_type(mtx->get_values())};
+}
+
+
+template <typename ValueType>
+kernels::cuda::compact_dense_wrapper<const kernels::cuda::cuda_type<ValueType>>
+compact(const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {kernels::cuda::as_cuda_type(mtx->get_const_values())};
+}
+
+
+template <typename ValueType>
+kernels::cuda::cuda_type<ValueType> *vector(matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return kernels::cuda::as_cuda_type(mtx->get_values());
+}
+
+
+template <typename ValueType>
+const kernels::cuda::cuda_type<ValueType> *vector(
+    const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return kernels::cuda::as_cuda_type(mtx->get_const_values());
+}
+
+
+template <typename KernelFunction, typename... KernelArgs>
+void run_kernel(std::shared_ptr<const CudaExecutor> exec, KernelFunction fn,
+                size_type size, KernelArgs &&... args)
+{
+    cuda::device_guard guard{exec->get_device_id()};
+    constexpr auto block_size = kernels::cuda::default_block_size;
+    auto num_blocks = ceildiv(size, block_size);
+    kernels::cuda::generic_kernel_1d<<<num_blocks, block_size>>>(
+        size, fn, kernels::cuda::map_to_device(args)...);
```

Comment:
Not having access to the size of the array prevents certain implementations. For example, if I want to initialize a reverse-counter (counting backwards from N) or want to do a bound-check myself, I would have to forward the size twice (once for the `generic_kernel`, the other time for my own lambda `fn`). If I had access to the size all the time, there would also be no need for the functions `compact` and `vector` as the author of the kernel could implement the accesses him-/herself (this would also make it obvious that it is an optimization, which `compact` or `vector` does not communicate).
Therefore, I would propose that all `KernelFunction`s get the size as the first argument, followed by the current index.
(same applies for 2D case.)

Reply:
I think what you are suggesting here would exceed the intended purpose of the framework, as simple kernels like these should not need to do any bounds checks, and would additionally add 2 unused parameters to all kernels. If the implementations do need access to the size, we can pass it explicitly as a parameter.

Reply:
It would be only a single (mostly useless) parameter (in case of `dim<N>` usage), but I understand your point.